### PR TITLE
Preserve data

### PR DESCRIPTION
**Pull request by nevisende**
# Preserve data
 - We checked the documentation for Ruby's [JSON library](https://ruby-doc.org/stdlib-3.0.0/libdoc/json/rdoc/JSON.html)[](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_preserve_data.md#need-a-big-picture).
 - On exit from our app is saved all the data (books, people, and rentals) into JSON files. One per type of data we are storing and named accordingly (e.g. `rentals.json`)
- On startup of our app is loaded all data from files. Handled the case when one or more of the files are missing.

### CODE REVIEW

File: store.rb [hunk 1]
```diff
@@ -1,31 +1,45 @@
...
+class Store
+  def self.read_convert(filename)
+    JSON.parse(File.read(filename))
   end
 
-  def push(item)
-    item.is_a?(Person) ? @person_arr.push(item) : @book_arr.push(item)
+  def self.push(item)
+    if item.is_a?(Person)
+      json_to_arr = read_convert('people.json')
```
Comment (TedLivist) [inline store.rb:10]: - [ ] When I delete your `.json` storage files, I get errors (see the first image) when trying to store a created **Person**. As stated in the requirements (see the second image), your app should handle cases when the storage files are not available. Kindly ensure that a new `people.json` file is created when not available.

  **SUGGESTION:** Just after the `if` statement here, you can create a `people.json` file if it does not exist like so:
  ```
  if item.is_a?(Person)
    File.open('people.json', 'w') {|f| f.write JSON.generate([]) } unless File.exist? 'people.json'
    ....
  ```


  Your error | The requirements
  :---:|:---:
  anselem-people-json | anselem-json-req

File: store.rb [hunk 2]
```diff
@@ -1,31 +1,45 @@
...
-  def list_all_books
-    @book_arr.map { |book| puts "Title: \"#{book.title}\" Author: #{book.author}" }
+  def self.list_all_books
+    JSON.parse(File.read('books.json')).map { |book| puts "Title: \"#{book['title']}\" Author: #{book['author']}" }
   end
 
-  def list_all_people
-    @person_arr.map { |person| puts "[#{person.type}] Name: #{person.name}, ID: #{person.id}, Age: #{person.age}" }
+  def self.list_all_people
+    JSON.parse(File.read('people.json')).map do |person|
```
Comment (TedLivist) [inline store.rb:30]: - [ ] You should also handle the missing storage `people.json` file here if the file does not exist.

  **SUGGESTION:** After the method declaration, you can return an empty array if the file does not exist, like so: 
  ```
  def self.list_all_people
    return [] unless File.exists?('people.json')
    
    JSON.parse(File.read...
    ...
  ```
  anselem-people-json-2

File: store.rb [hunk 3]
```diff
@@ -1,31 +1,45 @@
...
+      if item.type == 'Student'
+        json_to_arr.push({ type: item.type, 'name' => item.name, 'id' => item.id, 'age' => item.age,
+                           'parent_permission' => item.parent_permission, 'rentals' => item.rentals })
+      else
+        json_to_arr.push({ type: item.type, 'name' => item.name, 'id' => item.id, 'age' => item.age,
+                           'specialization' => item.specialization, 'rentals' => item.rentals })
+      end
+      File.write('people.json', JSON.generate(json_to_arr))
+    else
+      json_to_arr = read_convert('books.json').push({ 'title' => item.title, 'author' => item.author })
```
Comment (TedLivist) [inline store.rb:20]: - [ ] Kindly ensure that a new `books.json` file is created when not available.

File: store.rb [hunk 4]
```diff
@@ -1,31 +1,45 @@
...
+      File.write('books.json', JSON.generate(json_to_arr))
+    end
   end
 
-  def list_all_books
-    @book_arr.map { |book| puts "Title: \"#{book.title}\" Author: #{book.author}" }
+  def self.list_all_books
+    JSON.parse(File.read('books.json')).map { |book| puts "Title: \"#{book['title']}\" Author: #{book['author']}" }
```
Comment (TedLivist) [inline store.rb:26]: - [ ] Kindly ensure that the missing `books.json` file is handled when not available as well.

File: creation.rb [hunk 5]
```diff
@@ -31,58 +32,76 @@ def self.create_student(store)
...
+        person['rentals'] << { 'date' => rental.date, 'author' => rental.book['author'],
+                               'title' => rental.book['title'] }
+      end
+    end
+    File.write('people.json', JSON.generate(new_people_arr))
+    json_to_arr = rentals_arr.push({ 'date' => rental.date, 'person' => rental.person,
+                                     'book' => rental.book })
+    File.write('rentals.json', JSON.generate(json_to_arr))
     success('Rental')
   end
```
Comment (TedLivist) [inline creation.rb:106]: - [ ] Kindly ensure that the `rentals.json` file is created when it is not available.

  anselem-rentals

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @nevisende and @AnselemOdims,

Good job so far!
✔️ Implementation of storage files
✔️ Data is stored when files are present
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
### 🛑 Check this out too 🛑 
- [x] **In your README, kindly add instructions on how to run the app in locally (`ruby main.rb`)**

- [ ] _Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: creation.rb (hunk 5 above)
Comment (nevisende) [reply creation.rb]: Thank you for your review and suggestion :)

**Pull request merged**